packages/runtime/container-runtime/src/dataStores.ts

```diff
@@ -245,4 +247,80 @@ export class DataStores implements IDisposable {
 
     public get disposed() {return this.disposeOnce.evaluated;}
     public readonly dispose = () => this.disposeOnce.value;
+
+    public updateLeader() {
+        for (const [, context] of this.contexts) {
+            context.updateLeader(this.runtime.leader);
+        }
+    }
+
+    public resubmitDataStoreOp(content: any, localOpMetadata: unknown) {
+        const envelope = content as IEnvelope;
+        const context = this.contexts.get(envelope.address);
+        assert(!!context, "There should be a store context for the op");
+        context.reSubmit(envelope.contents, localOpMetadata);
+    }
+
+    public processFluidDataStoreOp(message: ISequencedDocumentMessage, local: boolean, localMessageMetadata: unknown) {
+        const envelope = message.contents as IEnvelope;
+        const transformed = { ...message, contents: envelope.contents };
+        const context = this.contexts.get(envelope.address);
+        assert(!!context, "There should be a store context for the op");
+        context.process(transformed, local, localMessageMetadata);
+    }
+
+    public async getDataStore(id: string, wait: boolean): Promise<IFluidDataStoreChannel> {
+        const deferredContext = this.contexts.prepDeferredContext(id);
+
+        if (!wait && !deferredContext.isCompleted) {
+            throw new Error(`DataStore ${id} does not exist`);
+        }
+
+        const context = await deferredContext.promise;
+        return context.realize();
+    }
+
+    public processSignal(address: string, message: IInboundSignalMessage, local: boolean) {
+        const context = this.contexts.get(address);
+        if (!context) {
+            // Attach message may not have been processed yet
+            assert(!local);
```

Comment:
nit: Consider adding a message to the assert which will make debugging slightly easier.

Reply:
just copy pasting for now :)
```suggestion
            assert(!local, "Missing datastore for local signal");
```

Reply:
It just caught me eye :-).